Analysis/bin/ShuffleMergeSpectral.cxx

```diff
@@ -0,0 +1,773 @@
+    run::Argument<std::string> tau_ratio{"tau-ratio", "ratio of tau types in the final spectrum \
+                                                      (if to take all taus of type e => e:-1 )"};
+    run::Argument<double> start_entry{"start-entry", "starting ratio from which file will be processed", 0};
+    run::Argument<double> end_entry{"end-entry", "end ratio until which file will be processed", 1};
+    run::Argument<double> exp_disbalance{"exp-disbalance", "maximal expected disbalance between low pt and high pt regions",0};
+
+};
+
+struct SourceDesc {
+    using Tau = tau_tuple::Tau;
+    using TauType = analysis::TauType;
+    using TauTuple = tau_tuple::TauTuple;
+    using SampleType = analysis::SampleType;
+    using Generator = std::mt19937_64;
+
+    SourceDesc(const std::string& name_, const Int_t&  _group_hash,
+               const std::vector<std::string>& _file_names, const std::vector<Int_t>& _name_hashes,
+               const std::set<std::string>& _disabled_branches, const double& _begin_rel, const double& _end_rel,
+               const std::set<TauType>& _tautypes, Generator& _gen) :
+        name(name_),  group_hash(_group_hash),
+        disabled_branches(_disabled_branches), entry_begin_rel(_begin_rel), entry_end_rel(_end_rel),
+        tau_types(_tautypes), current_n_processed(0), files_n_total(_file_names.size()),
+        entries_end(std::numeric_limits<size_t>::infinity()), total_n_processed(0), gen(&_gen)
+    {
+        if(_file_names.size()!=_name_hashes.size())
+          throw analysis::exception("file_names and names vectors have different size.");
+        if(_file_names.empty())
+          throw analysis::exception("Empty list of files for the source '%1%'.") % name;
+        if(!files_n_total)
+          throw analysis::exception("Empty source '%1%'.") % name;
+        ShuffleFiles(_file_names, _name_hashes);
+    }
+
+    SourceDesc(const SourceDesc&) = delete;
+    SourceDesc& operator=(const SourceDesc&) = delete;
+
+    bool DoNextStep()
+    {
+      do {
+        if(current_file_index == files_n_total-1 && current_n_processed >= entries_end) return false;
+        while(!current_file_index || current_n_processed == entries_end) {
+            if(!current_file_index)
+                current_file_index = 0;
+            else
+                ++(*current_file_index);
+            if(*current_file_index >= file_names.size())
+                throw analysis::exception("The expected number of events = %1% is bigger than the actual number of"
+                                          " end point events in source '%2%'.") % entries_file % name;
+            const std::string& file_name = file_names.at(*current_file_index);
+            dataset_hash = dataset_hash_arr.at(*current_file_index);
+            current_tuple.reset();
+            current_file = root_ext::OpenRootFile(file_name);
+            current_tuple = std::make_shared<TauTuple>("taus", current_file.get(), true, disabled_branches);
```

Comment:
@kandrosov small question: does it grantee that memory allocated for current_file and current_tuple is cleaned properly after each iteration?

The problem with huge virtual memory consumption was solved by switching to the "SourceDesc per data group" concept, but I still do not understand why RES still incises consumption after each iteration, it is not much, but it linearly increases. If I just move this 3 lines to the constructor of SourceDesc (open for one file always) - that is fine (it is stable).

Reply:
If the smart pointer is reset, the corresponding destructors must be called. In [SmartTree destructor](https://github.com/cms-tau-pog/TauMLTools/blob/master/Core/interface/SmartTree.h#L222-L223) the TTree is deleted. So if there are some memory leaks, it should be in the root code (as far as I can see)... You might try to add these two lines before opening a new file:
```c++
if(current_file)
    current_file->Close();
```
Beyond that, I don't know what else can we do reduce the leaks...